osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/MigratePlaylistScoresToSoloScoresCommand.cs

```diff
@@ -0,0 +1,190 @@
+
+                        MultiplayerScore[] scores = (await db.QueryAsync<MultiplayerScore>($"SELECT * FROM multiplayer_scores WHERE playlist_item_id = {item.id}")).ToArray();
+
+                        foreach (var score in scores)
+                        {
+                            Console.WriteLine($"Reprocessing score {score.id} from playlist {item.id}");
+
+                            Ruleset ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(item.ruleset_id);
+                            HitResult maxRulesetJudgement = ruleset.GetHitResults().First().result;
+                            Dictionary<HitResult, int> statistics = JsonConvert.DeserializeObject<Dictionary<HitResult, int>>(score.statistics)!;
+                            List<HitResult> allHits = statistics
+                                                      .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value))
+                                                      .ToList();
+                            var maximumStatistics = new Dictionary<HitResult, int>();
+
+                            foreach (var groupedStats in allHits
+                                                         .Select(r => getMaxJudgementFor(r, maxRulesetJudgement))
+                                                         .GroupBy(r => r))
+                            {
+                                maximumStatistics[groupedStats.Key] = groupedStats.Count();
+                            }
+
+                            APIMod[] roomMods = JsonConvert.DeserializeObject<APIMod[]>(item.required_mods)!;
+                            APIMod[] scoreMods = JsonConvert.DeserializeObject<APIMod[]>(score.mods)!;
+
+                            foreach (var m in roomMods)
+                                Debug.Assert(scoreMods.Contains(m));
+
+                            Console.WriteLine($"Mods: {string.Join(',', scoreMods.Select(m => m.ToString()))}");
+
+                            int insertId = await db.InsertAsync(new SoloScore
+                            {
+                                user_id = (int)score.userId,
+                                beatmap_id = (int)score.beatmapId,
+                                ruleset_id = item.ruleset_id,
+                                preserve = true,
+                                ScoreInfo = SoloScoreInfo.ForSubmission(new ScoreInfo
+                                {
+                                    Statistics = statistics,
+                                    MaximumStatistics = maximumStatistics,
+                                    Ruleset = ruleset.RulesetInfo,
+                                    MaxCombo = score.max_combo,
+                                    APIMods = scoreMods,
+                                    Accuracy = (double)score.accuracy!,
+                                    TotalScore = (long)score.total_score!,
+                                    Rank = Enum.Parse<ScoreRank>(score.rank),
```

Comment:
All of these are nullable in `multiplayer_scores`, so any of these could die in practice. Could probably happen if a client dies before submission or something. I definitely have such rows on my test db, although I cannot recall specific circumstances. Another thing to check for before attempting to run I guess?

Reply:
right, I forgot about the non-finished score case. Those should probably still have score_link entry for attempts count but no solo_score entry.